src/main.rs

```diff
@@ -294,10 +311,23 @@ async fn main() -> Result<()> {
         .parse()?;
     let enable_audio = matches.is_present("enable-audio")
         || std::env::var("VNC_ENABLE_EXPERIMENTAL_AUDIO").unwrap_or_else(|_| String::new()) != "";
-    let authentication = match enable_audio {
-        false => Arc::new(auth::RfbAuthentication::Null),
-        true => Arc::new(auth::RfbAuthentication::Passthrough),
+    let authentication = if matches.value_of("replid").is_some() {
+        let mut pubkeys_base64: HashMap<String, String> =
+            serde_json::from_str(matches.value_of("pubkeys").unwrap())?;
+        let pubkeys = pubkeys_base64
+            .drain()
+            .map(|(keyid, pubkey)| Ok((keyid, base64::decode(pubkey)?)))
+            .collect::<std::result::Result<HashMap<String, Vec<u8>>, base64::DecodeError>>()?;
+        Arc::new(auth::RfbAuthentication::Replit {
+            replid: matches.value_of("replid").unwrap().to_string(),
+            pubkeys,
+        })
+    } else if enable_audio {
```

Comment:
I didn't ask this before, but why do we switch authentications methods based on whether or not audio is enabled?

Reply:
oh, good question! when audio is not enabled we go the easiest possible route: blindly copying bytes between the WebSocket <-> Socket: https://github.com/replit/rfbproxy/blob/3ca838f1f4a2127f85f602374550282440e9d340/src/main.rs#L31 . when we take that route, we don't need to do anything with the stream, we let both parties talk directly to each other.

When audio is enabled, OTOH, this process needs to be aware of the message boundaries to be able to inject the audio messages in a place where it won't cause the other party to de-synchronize and terminate the connection (RFB is unfortunately not self-delimited and is _very_ stateful: the length of each message depends on the message type and sometimes some shared state between the parties). So when audio is enabled, we need to leave the stream in a well-known state so that we can continue parsing messages one at a time.

there is a comment to that regards elsewhere, but it doesn't hurt to repeat it here to avoid reader surprises.

Reply:
Neat, makes sense!